src/psyclone/domain/lfric/algorithm/psyir_to_algpsyir.py

```diff
@@ -0,0 +1,178 @@
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author R. W. Ford STFC Daresbury Lab
+
+'''Specialise generic PSyIR representing an algorithm layer to an
+LFRic algorithm-layer-specific PSyIR which uses specialised classes.
+
+'''
+from fparser.two.Fortran2003 import \
+    Actual_Arg_Spec, Name, Char_Literal_Constant, Structure_Constructor
+from psyclone.psyir.nodes import Call, CodeBlock, ArrayReference
+from psyclone.psyir.symbols import Symbol, TypeSymbol, \
+    StructureType, RoutineSymbol
+from psyclone.domain.lfric.algorithm import \
+    LfricBuiltinRef, LfricCodedKernelRef, LfricAlgorithmInvokeCall
+from psyclone.domain.lfric.lfric_builtins import BUILTIN_MAP as builtins
+from psyclone.errors import GenerationError, InternalError
+from psyclone.psyir.frontend.fparser2 import Fparser2Reader
+
+
+def psyir_to_algpsyir(psyir):
+    '''Takes a generic PSyIR tree and translates it to an
+    LFRic algorithm-specific PSyIR representation
+
+    :param psyir: generic-PSyIR tree.
+    :type psyir: subclass of :py:class:`psyclone.psyir.nodes.node`
+
+    :raises GenerationError: if a PSyIR CodeBlock represents an \
+        argument and the CodeBlock contains more than one statement (as \
+        only one is expected).
+    :raises GenerationError: if more than one named argument is \
+        provided in an invoke call.
+    :raises GenerationError: if a named argument has an incorrect \
+        format.
+    :raises InternalError: if an unexpected argument type is found.
+
+    '''
+    for call in psyir.walk(Call):
```

Comment:
This has some parts repeated from the generic psyir_to_algpsyir. I see a few ways to avoid this e.g. being a class with a __call__ method and an a private method to do the API-specific  part inside `for call_arg in call.children`, but before doing any modification, is there any particular reason you chose a function here?

In some sense if feels like a transformation, would it be possible to encapsulate this as a PSyclone transformation (which can also use OO inheritance to don't repeat the common parts)? What are your thoughts on this?

Reply:
As discussed, I've re-written as transformations and tried to re-use where possible.